Log decoding

nathanleiby opened the pull request:
https://clever.atlassian.net/browse/INFRA-2228

Parses log and adds additional fields (`ParseAndEnhance`). This is to get parity with how we decode logs in our existing Heka setup.

nathanleiby committed:
decode: first decoders-- Rsyslog, Kayvee

nathanleiby committed:
glide: update deps

nathanleiby committed:
decode: FieldsFromHaproxy

nathanleiby committed:
decode: full parsing, including contaienrMeta

nathanleiby committed:
decode: add Titles to tests

nathanleiby committed:
decode: allow override to container_{env,app,task}

nathanleiby committed:
decode: cleanup

nathanleiby committed:
fully decode the message into fields

also allows configuring 'env' field in logs, by setting DEPLOY_ENV env var

mohit commented on README.md:
```diff
@@ -19,6 +19,7 @@ FIREHOSE_AWS_REGION=us-west-2 \
 FIREHOSE_STREAM_NAME=firehose-test \
 LOG_FILE=/tmp/kcl_stderr \
 RATE_LIMIT=100 \
+DEPLOY_ENV="local" \
```
This should be added by ark now automatically right?

nathanleiby replied:
`_DEPLOY_ENV` is injected by Catapult -- It should be equivalent to use that instead, it just puts the burden on Catapult instead of the app. 

In the past I think we've been uncertain as to whether app code should depend on Catapult injected env vars.

nathanleiby replied:
The cases where I think we've agreed it's OK is when the app includes an infra library (kayvee), so from the app's perspective it's unaware of how those env vars are used.

nathanleiby commented:
AFAIK, this recreates all of the decoding and field injecting logic that we had from Heka except for:

- parsing Haproxy logs. This seems low priority, and we also could consider switching standalone apps that output Haproxy logs to output in JSON format instead. We also could use https://github.com/satyrius/gonx to parse.
- adding KV routing. This isn't necessary for the log archive, so we can add it when needed for another consumer.

nathanleiby commented:
cc @xavi-

mohit commented on cmd/consumer/main.go:
```diff
@@ -30,12 +30,13 @@ func main() {
 
 	sess := session.Must(session.NewSession(aws.NewConfig().WithRegion(getEnv("FIREHOSE_AWS_REGION")).WithMaxRetries(4)))
```
why not use the default `AWS_REGION` which is automatically read by the aws-sdk?

nathanleiby replied:
Our setup depends on reading from one region and writing to another.

nathanleiby committed:
decode: programname should have 12 char AWS accountID

nathanleiby committed:
decode: remove 'logtag' field, unused

mohit commented on decode/decode.go:
```diff
@@ -0,0 +1,196 @@
+// - RSYSLOG_FileFormat - a modern-style logfile format similar to TraditionalFileFormat, but with high-precision timestamps and timezone information
+//
+// For more details on Rsylog formats: https://rsyslog-5-8-6-doc.neocities.org/rsyslog_conf_templates.html
+func FieldsFromSyslog(line string) (map[string]interface{}, error) {
```
we should write a go bench for this function.

nathanleiby replied:
Added benchmarks, and they'll be written as Circle artifacts so we can look over time.

8504fa8

nathanleiby committed:
decoder: get rid of `Payload`. only use `rawlog`

Payload was a field that had special meaning in Heka, but practically:

1. if it has data, it's exactly the same data as in `rawlog`
2. sometimes, it doesn't have data (since it gets removed by some Heka
decoders/filters)

So let's *only* have `rawlog`, and get rid of Payload.

mohit reviewed (commented):
Looks good. We should add some bench to the important functions that get hit with each log line just for tracking over time.

nathanleiby committed:
decode: simplify error type check. add test

nathanleiby committed:
Makefile: actually run tests

nathanleiby committed:
message_batcher: fix data race. (also linting)

nathanleiby committed:
decode: add benchmarks

xavi- commented on decode/decode.go:
```diff
@@ -0,0 +1,194 @@
+
+	m["Type"] = "Kayvee"
+
+	return m, nil
```
Not sure what the right answer is, but something to consider is that the Kayvee spec says the a `source` and `title` should be specified.  Perhaps we should enforce that in the decoder and in the struct that returned?

nathanleiby replied:
Since we were just using a JSON decoder before and calling it Kayvee, I think it's reasonable to not yet make it strict.

https://github.com/Clever/heka-clever-plugins/blob/master/lua/decoders/json_decoder.lua

If we need to distinguish between Kayvee and JSON, we can add that logic later

nathanleiby committed:
circle: deploy master builds

nathanleiby committed:
glide: update deps

nathanleiby committed:
linting fix

nathanleiby committed:
Makefile: remove go version check (breaks Docker build)

Since the Docker image relies on the Makefile, but it doesn't have Go
installed, this check breaks it. Since it's pretty low value, let's
remove.
